ocaml/xapi/livepatch.ml

```diff
@@ -0,0 +1,279 @@
+        fail "Not found build id"
+    | false -> (
+        LE.any_int32 >>= fun name_size ->
+        LE.any_int32 >>= fun desc_size ->
+        LE.any_int32 >>= fun note_type ->
+        take (align name_size) >>= fun name ->
+        take (align desc_size) >>= fun desc ->
+        let name' = Astring.(String.trim ~drop:Char.Ascii.is_control name) in
+        match (note_type, desc_size, name') with
+        | 3l, 20l, "GNU" ->
+            return desc
+        | _ ->
+            section' ()
+      )
+
+  let section = section' ()
+
+  let get_base_build_id () =
+    let open Rresult.R.Infix in
+    let consume = Angstrom.Consume.Prefix in
+    let path = "/sys/kernel/notes" in
+    let ic = open_in path in
+    Fun.protect ~finally:(fun () -> close_in ic) @@ fun () ->
+    let size = in_channel_length ic in
+    let r =
+      Angstrom.(parse_string ~consume section) (really_input_string ic size)
+      >>= fun build_id_bs ->
+      Angstrom.(parse_string ~consume BuildId.all_bytes build_id_bs)
+    in
+    match r with
+    | Ok ss ->
+        let build_id = Astring.String.concat ~sep:"" ss in
+        debug "Kernel build id: %s" build_id ;
+        Some build_id
+    | Error msg ->
+        warn "Can't get kernel build id: %s" msg ;
+        None
+end
+
+module XenLivePatch = struct
+  let get_running_livepatch () =
+    let r = Re.Posix.compile_pat {|^[ ]*lp_([^_ ]+)_([^_ ]+).+APPLIED.*$|} in
+    Helpers.call_script !Xapi_globs.xen_livepatch_cmd ["list"]
+    |> Astring.String.cuts ~sep:"\n"
+    |> List.filter_map (fun line ->
+           match Re.exec_opt r line with
+           | Some groups -> (
+               let base_vr = Re.Group.get groups 1 in
+               let to_vr = Re.Group.get groups 2 in
+               match
+                 Astring.String.(cuts ~sep:"-" base_vr, cuts ~sep:"-" to_vr)
+               with
+               | [b_v; b_r], [t_v; t_r] ->
+                   Some (b_v, b_r, t_v, t_r)
+               | _ ->
+                   None
+             )
+           | None ->
+               None
+       )
+    |> get_latest_livepatch
+
+  let get_base_build_id () =
+    let drop x =
+      let open Astring.Char.Ascii in
+      is_control x || is_blank x || is_white x
+    in
+    Helpers.call_script !Xapi_globs.xl_cmd ["info"; "build_id"]
+    |> Astring.String.trim ~drop
+    |> function
+    | build_id when not (Astring.String.is_empty build_id) ->
+        debug "Xen build id: %s" build_id ;
+        Some build_id
+    | _ ->
+        warn "Can't get Xen build id" ;
+        None
+end
+
+let get_livepatch_dir ~component ~base_build_id =
+  let comp_str = string_of_component component in
+  Printf.sprintf "/usr/lib/%s-livepatch/%s" comp_str base_build_id
+
+let get_installed_livepatch_file ~component ~base_build_id =
+  let lp_dir = get_livepatch_dir ~component ~base_build_id in
+  let installed_symlink = Filename.concat lp_dir "livepatch.livepatch" in
+  if not (Sys.file_exists installed_symlink) then
```

Comment:
This could be done with a single match: 
```
match Unix.lstat installed_symlink with
| Unix.{st_kind = S_LNK;_} -> ... Option .some
| exception _ -> None
```